google/cloud/forseti/services/inventory/crawler.py

```diff
@@ -116,15 +135,110 @@ def update(self, resource):
             raise
 
 
+class ParallelCrawler(Crawler):
+    """Multi-threaded Crawler implementation."""
+
+    def __init__(self, config):
+        super(ParallelCrawler, self).__init__(config)
+        self._write_lock = threading.Lock()
+        self._dispatch_queue = Queue()
+        self._shutdown_event = threading.Event()
+
+    def _start_workers(self):
+        """Start a pool of worker threads for processing the dispatch queue."""
+        self._shutdown_event.clear()
+        for _ in xrange(self.config.threads):
+            worker = threading.Thread(target=self._process_queue)
+            worker.daemon = True
+            worker.start()
+
+    def _process_queue(self):
+        """Process items in the queue until the shutdown event is set."""
+        while not self._shutdown_event.is_set():
+            try:
+                (callback, resource) = self._dispatch_queue.get(timeout=1)
+            except Empty:
+                continue
+
+            try:
+                callback()
+            except Exception as e:
```

Comment:
Is it possible to have this handler in the callback?

The reason I'm asking is that I believe what the handler does is a implementation detail of the resource so it should be part of the callback closure. This way, the crawler just schedules callables and does really not have to care what these callables do.

Reply:
Fixed.